stage0/src/apic.rs

```diff
@@ -0,0 +1,249 @@
+//
+// Copyright 2023 The Project Oak Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+
+use core::arch::x86_64::__cpuid;
+
+use bitflags::bitflags;
+use x86_64::{registers::model_specific::Msr, PhysAddr};
+
+bitflags! {
+    #[derive(Clone, Copy, Debug)]
+    pub struct ApicBaseFlags: u64 {
+        /// APIC Enable
+        const AE = (1 << 11);
+
+        /// x2APIC Mode Enable
+        const EXTD = (1 << 10);
+
+        /// Boot Strap CPU Core
+        const BSC = (1 << 8);
+    }
+}
+
+/// The APIC Base Address Register.
+pub struct ApicBase;
+
+impl ApicBase {
+    pub const MSR: Msr = Msr::new(0x0000_001B);
+
+    fn read_raw() -> u64 {
+        unsafe { Self::MSR.read() }
+    }
+
+    fn write_raw(value: u64) {
+        let mut msr = Self::MSR;
+        unsafe { msr.write(value) }
+    }
+
+    pub fn read() -> (PhysAddr, ApicBaseFlags) {
+        let val = Self::read_raw();
+        let apa = PhysAddr::new(val & 0x000F_FFFF_FFFF_F000u64);
+        let flags = ApicBaseFlags::from_bits_truncate(val);
+
+        (apa, flags)
+    }
+
+    pub fn write(apa: PhysAddr, flags: ApicBaseFlags) {
+        Self::write_raw(flags.bits() | apa.as_u64());
+    }
+}
+
+pub struct X2APICIdRegister;
+
+impl X2APICIdRegister {
+    pub const MSR: Msr = Msr::new(0x0000_0802);
+
+    fn read_raw() -> u64 {
+        unsafe { Self::MSR.read() }
+    }
+
+    pub fn read() -> u32 {
+        Self::read_raw() as u32
+    }
+}
+
+pub struct InterruptCommandRegister;
+
+#[allow(dead_code)]
+pub enum MessageType {
+    /// IPI delivers an interrupt to the target local APIC specified in the Destination field.
+    Fixed = 0b000,
+
+    /// IPI delivers an SMI interrupt to the target local APIC(s). Trigger mode is edge-triggered
+    /// and Vector must be 0x00.
+    SMI = 0b010,
```

Comment:
nit: clippy doesn't like the casing here, nor NMI below.

Reply:
SMI and NMI are acronyms that I think should be capitalized, so I've told Clippy to shut up here.

Reply:
based imo